### PR TITLE
AEA-1135 Validation of protocol specification

### PR DESCRIPTION
**Pull request by 5A11**
This PR streamlines and adds more validation for protocol specification, especially for the dialogue section.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/configurations/base.py [hunk 1]
```diff
@@ -2045,20 +2045,6 @@ def __init__(self, **args):
         """Initialize a speech_act content configuration."""
         super().__init__()
         self.args = args  # type: Dict[str, str]
-        self._check_consistency()
-
-    def _check_consistency(self):
-        """Check consistency of the args."""
-        for content_name, content_type in self.args.items():
-            if not isinstance(content_name, str) or not isinstance(content_type, str):
```
Comment (5A11) [inline aea/configurations/base.py:2053]: The first part of the if was inconsequential/unreachable (if not string, yaml converter would raise error.)
The seance part added to the validator in the generator.

File: aea/configurations/base.py [hunk 2]
```diff
@@ -2045,20 +2045,6 @@ def __init__(self, **args):
...
-                raise ProtocolSpecificationParseError(
-                    "Contents' names and types must be string."
-                )
-            # Check each content definition key/value (i.e. content name/type) is not empty
-            if content_name == "" or content_type == "":
```
Comment (5A11) [inline aea/configurations/base.py:2058]: Already captured in the regex check in the generator validation.

File: aea/configurations/base.py [hunk 3]
```diff
@@ -2153,35 +2139,8 @@ def from_json(cls, obj: Dict):
             protocol_specification.speech_acts.create(
                 speech_act, speech_act_content_config
             )
-        protocol_specification._check_consistency()  # pylint: disable=protected-access
         return protocol_specification
 
-    def _check_consistency(self):
-        """Validate the correctness of the speech_acts."""
-        if len(self.speech_acts.read_all()) == 0:
```
Comment (5A11) [inline aea/configurations/base.py:2161]: Added to the generator's validation.

File: aea/configurations/base.py [hunk 4]
```diff
@@ -2153,35 +2139,8 @@ def from_json(cls, obj: Dict):
...
-            raise ProtocolSpecificationParseError(
-                "There should be at least one performative defined in the speech_acts."
-            )
-        content_dict = {}
-        for performative, speech_act_content_config in self.speech_acts.read_all():
-            if not isinstance(performative, str):
```
Comment (5A11) [inline aea/configurations/base.py:2167]: unreachable (yaml converter would raise if it is the case).

File: aea/configurations/base.py [hunk 5]
```diff
@@ -2153,35 +2139,8 @@ def from_json(cls, obj: Dict):
...
-                raise ProtocolSpecificationParseError(
-                    "A 'performative' is not specified as a string."
-                )
-            if performative == "":
-                raise ProtocolSpecificationParseError(
```
Comment (5A11) [inline aea/configurations/base.py:2172]: Already captured by the regex validation in the generator.

File: aea/configurations/base.py [hunk 6]
```diff
@@ -2153,35 +2139,8 @@ def from_json(cls, obj: Dict):
...
-                    "A 'performative' cannot be an empty string."
-                )
-            for content_name, content_type in speech_act_content_config.args.items():
```
Comment (5A11) [inline aea/configurations/base.py:2175]: Moved to generator's validation.

File: aea/protocols/generator/validate.py [hunk 7]
```diff
@@ -473,6 +522,30 @@ def _validate_protocol_buffer_schema_code_snippets(
...
+            return (
+                False,
+                "Missing required field '{}' in the dialogue section of the protocol specification.".format(
+                    required_field
+                ),
+            )
+
+    return (
+        True,
+        "Dialogue section has all the required fields and their values have correct types.",
```
Comment (DavidMinarsch) [inline aea/protocols/generator/validate.py:545]: Here types aren't checked. Fix string

**Review by DavidMinarsch — COMMENTED** [review]
LGTM, please make sure tests cover all cases

File: aea/protocols/generator/validate.py (hunk 7 above)
Comment (5A11) [reply aea/protocols/generator/validate.py:545]: Good eyes! Done.

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**